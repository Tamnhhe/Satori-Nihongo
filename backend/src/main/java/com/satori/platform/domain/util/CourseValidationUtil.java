package com.satori.platform.domain.util;

import com.satori.platform.domain.Course;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

/**
 * Utility class for Course validation logic.
 */
public final class CourseValidationUtil {

    private CourseValidationUtil() {
        // Utility class
    }

    /**
     * Validates a course and returns a list of validation errors.
     * 
     * @param course the course to validate
     * @return list of validation error messages
     */
    public static List<String> validateCourse(Course course) {
        List<String> errors = new ArrayList<>();

        if (course == null) {
            errors.add("Course cannot be null");
            return errors;
        }

        // Title validation
        if (course.getTitle() == null || course.getTitle().trim().isEmpty()) {
            errors.add("Course title is required");
        } else if (course.getTitle().length() > 255) {
            errors.add("Course title cannot exceed 255 characters");
        }

        // Course code validation
        if (course.getCourseCode() != null && course.getCourseCode().length() > 20) {
            errors.add("Course code cannot exceed 20 characters");
        }

        // Description validation
        if (course.getDescription() != null && course.getDescription().length() > 2000) {
            errors.add("Course description cannot exceed 2000 characters");
        }

        // Duration validation
        if (course.getDuration() != null && course.getDuration() <= 0) {
            errors.add("Course duration must be positive");
        }

        if (course.getEstimatedDuration() != null && course.getEstimatedDuration() <= 0) {
            errors.add("Estimated duration must be positive");
        }

        // Price validation
        if (course.getPrice() != null && course.getPrice().compareTo(BigDecimal.ZERO) < 0) {
            errors.add("Course price cannot be negative");
        }

        // Business logic validation
        if (course.getDuration() != null && course.getEstimatedDuration() != null) {
            if (course.getDuration() > course.getEstimatedDuration() * 3) {
                errors.add("Course duration seems too long compared to estimated duration");
            }
        }

        return errors;
    }

    /**
     * Checks if a course is valid for publication.
     * 
     * @param course the course to check
     * @return true if the course can be published
     */
    public static boolean isReadyForPublication(Course course) {
        if (course == null) {
            return false;
        }

        return course.getTitle() != null && !course.getTitle().trim().isEmpty() &&
                course.getDescription() != null && !course.getDescription().trim().isEmpty() &&
                course.getDifficultyLevel() != null &&
                course.getLanguage() != null &&
                course.getLessonCount() > 0;
    }

    /**
     * Generates a course code if not provided.
     * 
     * @param course the course
     * @return generated course code
     */
    public static String generateCourseCode(Course course) {
        if (course.getCourseCode() != null && !course.getCourseCode().trim().isEmpty()) {
            return course.getCourseCode();
        }

        StringBuilder code = new StringBuilder();

        // Add language prefix
        if (course.getLanguage() != null) {
            code.append(course.getLanguage().name().substring(0, 2));
        } else {
            code.append("GN"); // General
        }

        // Add difficulty level
        if (course.getDifficultyLevel() != null) {
            switch (course.getDifficultyLevel()) {
                case BEGINNER -> code.append("01");
                case ELEMENTARY -> code.append("02");
                case INTERMEDIATE -> code.append("03");
                case UPPER_INTERMEDIATE -> code.append("04");
                case ADVANCED -> code.append("05");
                case PROFICIENT -> code.append("06");
            }
        } else {
            code.append("00");
        }

        // Add timestamp suffix to ensure uniqueness
        code.append(System.currentTimeMillis() % 10000);

        return code.toString();
    }
}