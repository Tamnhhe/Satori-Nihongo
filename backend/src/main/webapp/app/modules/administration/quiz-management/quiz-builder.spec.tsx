import React from 'react';
import { screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';
import { QuizBuilder } from './quiz-builder';
import { renderWithProviders, createMockQuiz, expectToBeAccessible } from '../../../shared/util/test-utils';

expect.extend(toHaveNoViolations);

const mockQuestions = [
  {
    id: '1',
    content: 'What is the Japanese word for "hello"?',
    type: 'MULTIPLE_CHOICE',
    correctAnswer: 'こんにちは',
    imageUrl: null,
    suggestion: 'Think about common greetings',
    answerExplanation: 'こんにちは (konnichiwa) is the most common greeting in Japanese',
    options: ['こんにちは', 'さようなら', 'ありがとう', 'すみません'],
  },
  {
    id: '2',
    content: 'Fill in the blank: 私は___です。(I am a student)',
    type: 'FILL_IN_BLANK',
    correctAnswer: '学生',
    imageUrl: null,
    suggestion: 'What word means student in Japanese?',
    answerExplanation: '学生 (gakusei) means student in Japanese',
  },
];

const mockQuiz = createMockQuiz({
  id: '1',
  title: 'Basic Japanese Quiz',
  description: 'Test your basic Japanese knowledge',
  questions: mockQuestions,
});

const mockInitialState = {
  quizManagement: {
    loading: false,
    currentQuiz: mockQuiz,
    questions: mockQuestions,
    availableQuestions: mockQuestions,
    draggedQuestion: null,
    saving: false,
  },
  authentication: {
    account: {
      authorities: ['ROLE_ADMIN'],
      login: 'admin',
    },
  },
};

describe('QuizBuilder Component', () => {
  it('should render quiz builder with quiz details', () => {
    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    expect(screen.getByTestId('quiz-builder')).toBeInTheDocument();
    expect(screen.getByDisplayValue('Basic Japanese Quiz')).toBeInTheDocument();
    expect(screen.getByDisplayValue('Test your basic Japanese knowledge')).toBeInTheDocument();
  });

  it('should display quiz questions in order', () => {
    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    expect(screen.getByTestId('questions-list')).toBeInTheDocument();
    expect(screen.getByText('What is the Japanese word for "hello"?')).toBeInTheDocument();
    expect(screen.getByText(/Fill in the blank/)).toBeInTheDocument();
  });

  it('should handle quiz title changes', async () => {
    const onSave = jest.fn();
    const user = userEvent.setup();

    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={onSave} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    const titleInput = screen.getByTestId('quiz-title-input');
    await user.clear(titleInput);
    await user.type(titleInput, 'Updated Quiz Title');

    expect(titleInput).toHaveValue('Updated Quiz Title');
  });

  it('should handle quiz description changes', async () => {
    const user = userEvent.setup();

    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    const descriptionInput = screen.getByTestId('quiz-description-input');
    await user.clear(descriptionInput);
    await user.type(descriptionInput, 'Updated description');

    expect(descriptionInput).toHaveValue('Updated description');
  });

  it('should handle drag and drop question reordering', () => {
    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    const firstQuestion = screen.getByTestId('question-item-1');
    const secondQuestion = screen.getByTestId('question-item-2');

    // Simulate drag and drop
    fireEvent.dragStart(firstQuestion);
    fireEvent.dragOver(secondQuestion);
    fireEvent.drop(secondQuestion);

    expect(screen.getByTestId('questions-reordered')).toBeInTheDocument();
  });

  it('should add new question when add button is clicked', async () => {
    const onQuestionAdd = jest.fn();
    const user = userEvent.setup();

    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={onQuestionAdd} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    const addQuestionButton = screen.getByTestId('add-question-button');
    await user.click(addQuestionButton);

    expect(onQuestionAdd).toHaveBeenCalled();
  });

  it('should edit question when edit button is clicked', async () => {
    const onQuestionEdit = jest.fn();
    const user = userEvent.setup();

    renderWithProviders(
      <QuizBuilder
        quiz={mockQuiz}
        questions={mockQuestions}
        onSave={jest.fn()}
        onQuestionAdd={jest.fn()}
        onQuestionEdit={onQuestionEdit}
      />,
      { initialState: mockInitialState },
    );

    const editButtons = screen.getAllByTestId('edit-question-button');
    await user.click(editButtons[0]);

    expect(onQuestionEdit).toHaveBeenCalledWith(mockQuestions[0]);
  });

  it('should delete question when delete button is clicked', async () => {
    const user = userEvent.setup();

    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    const deleteButtons = screen.getAllByTestId('delete-question-button');
    await user.click(deleteButtons[0]);

    // Confirm deletion
    const confirmButton = screen.getByTestId('confirm-delete-button');
    await user.click(confirmButton);

    expect(screen.queryByText('What is the Japanese word for "hello"?')).not.toBeInTheDocument();
  });

  it('should show question types with proper icons', () => {
    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    expect(screen.getByTestId('question-type-MULTIPLE_CHOICE')).toBeInTheDocument();
    expect(screen.getByTestId('question-type-FILL_IN_BLANK')).toBeInTheDocument();
  });

  it('should handle quiz settings', async () => {
    const user = userEvent.setup();

    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    const settingsButton = screen.getByTestId('quiz-settings-button');
    await user.click(settingsButton);

    expect(screen.getByTestId('quiz-settings-modal')).toBeInTheDocument();

    const timeLimit = screen.getByTestId('time-limit-input');
    await user.clear(timeLimit);
    await user.type(timeLimit, '30');

    expect(timeLimit).toHaveValue('30');
  });

  it('should preview quiz when preview button is clicked', async () => {
    const user = userEvent.setup();

    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    const previewButton = screen.getByTestId('preview-quiz-button');
    await user.click(previewButton);

    expect(screen.getByTestId('quiz-preview-modal')).toBeInTheDocument();
  });

  it('should save quiz when save button is clicked', async () => {
    const onSave = jest.fn();
    const user = userEvent.setup();

    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={onSave} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    const saveButton = screen.getByTestId('save-quiz-button');
    await user.click(saveButton);

    await waitFor(() => {
      expect(onSave).toHaveBeenCalledWith(
        expect.objectContaining({
          title: 'Basic Japanese Quiz',
          description: 'Test your basic Japanese knowledge',
        }),
      );
    });
  });

  it('should validate quiz before saving', async () => {
    const onSave = jest.fn();
    const user = userEvent.setup();

    renderWithProviders(
      <QuizBuilder quiz={{ ...mockQuiz, title: '' }} questions={[]} onSave={onSave} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    const saveButton = screen.getByTestId('save-quiz-button');
    await user.click(saveButton);

    expect(screen.getByText('Quiz title is required')).toBeInTheDocument();
    expect(screen.getByText('At least one question is required')).toBeInTheDocument();
    expect(onSave).not.toHaveBeenCalled();
  });

  it('should show loading state when saving', () => {
    const savingState = {
      ...mockInitialState,
      quizManagement: {
        ...mockInitialState.quizManagement,
        saving: true,
      },
    };

    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: savingState },
    );

    expect(screen.getByTestId('saving-indicator')).toBeInTheDocument();
    expect(screen.getByTestId('save-quiz-button')).toBeDisabled();
  });

  it('should handle question bank integration', async () => {
    const user = userEvent.setup();

    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    const questionBankButton = screen.getByTestId('question-bank-button');
    await user.click(questionBankButton);

    expect(screen.getByTestId('question-bank-modal')).toBeInTheDocument();
  });

  it('should be accessible', async () => {
    const { container } = renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    await expectToBeAccessible(container);
  });

  it('should have proper ARIA labels and roles', () => {
    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    expect(screen.getByRole('main')).toHaveAttribute('aria-label', 'Quiz Builder');
    expect(screen.getByTestId('questions-list')).toHaveAttribute('role', 'list');

    const questionItems = screen.getAllByRole('listitem');
    expect(questionItems).toHaveLength(2);
  });

  it('should support keyboard navigation', async () => {
    const user = userEvent.setup();

    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    const titleInput = screen.getByTestId('quiz-title-input');
    titleInput.focus();

    await user.keyboard('{Tab}');
    expect(screen.getByTestId('quiz-description-input')).toHaveFocus();

    await user.keyboard('{Tab}');
    expect(screen.getByTestId('add-question-button')).toHaveFocus();
  });

  it('should handle undo/redo functionality', async () => {
    const user = userEvent.setup();

    renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    // Make a change
    const titleInput = screen.getByTestId('quiz-title-input');
    await user.clear(titleInput);
    await user.type(titleInput, 'New Title');

    // Undo
    await user.keyboard('{Control>}z{/Control}');
    expect(titleInput).toHaveValue('Basic Japanese Quiz');

    // Redo
    await user.keyboard('{Control>}{Shift>}z{/Shift}{/Control}');
    expect(titleInput).toHaveValue('New Title');
  });

  it('should match snapshot', () => {
    const { container } = renderWithProviders(
      <QuizBuilder quiz={mockQuiz} questions={mockQuestions} onSave={jest.fn()} onQuestionAdd={jest.fn()} onQuestionEdit={jest.fn()} />,
      { initialState: mockInitialState },
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
