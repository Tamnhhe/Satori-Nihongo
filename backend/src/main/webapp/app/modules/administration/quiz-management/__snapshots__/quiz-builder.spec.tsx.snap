// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizBuilder Component should match snapshot 1`] = `
<div
  style="padding: 24px;"
>
  <div
    style="display: flex; justify-content: space-between; align-items: center; margin-bottom: 24px;"
  >
    <h2>
      <span />
    </h2>
    <div
      class="ant-space css-dev-only-do-not-override-1vjf2v5 ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small"
    >
      <div
        class="ant-space-item"
      >
        <button
          class="ant-btn css-dev-only-do-not-override-1vjf2v5 ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
          type="button"
        >
          <span />
        </button>
      </div>
      <div
        class="ant-space-item"
      >
        <button
          class="ant-btn css-dev-only-do-not-override-1vjf2v5 ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
          type="button"
        >
          <span
            class="ant-btn-icon"
          >
            <span
              aria-label="save"
              class="anticon anticon-save"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="save"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M893.3 293.3L730.7 130.7c-7.5-7.5-16.7-13-26.7-16V112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V338.5c0-17-6.7-33.2-18.7-45.2zM384 184h256v104H384V184zm456 656H184V184h136v136c0 17.7 14.3 32 32 32h320c17.7 0 32-14.3 32-32V205.8l136 136V840zM512 442c-79.5 0-144 64.5-144 144s64.5 144 144 144 144-64.5 144-144-64.5-144-144-144zm0 224c-44.2 0-80-35.8-80-80s35.8-80 80-80 80 35.8 80 80-35.8 80-80 80z"
                />
              </svg>
            </span>
          </span>
          <span />
        </button>
      </div>
    </div>
  </div>
  <div
    class="ant-tabs ant-tabs-top css-dev-only-do-not-override-1vjf2v5"
  >
    <div
      aria-orientation="horizontal"
      class="ant-tabs-nav"
      role="tablist"
    >
      <div
        class="ant-tabs-nav-wrap"
      >
        <div
          class="ant-tabs-nav-list"
          style="transform: translate(0px, 0px);"
        >
          <div
            class="ant-tabs-tab ant-tabs-tab-active"
            data-node-key="basic"
          >
            <div
              aria-controls="rc-tabs-test-panel-basic"
              aria-selected="true"
              class="ant-tabs-tab-btn"
              id="rc-tabs-test-tab-basic"
              role="tab"
              tabindex="0"
            >
              Basic Information
            </div>
          </div>
          <div
            class="ant-tabs-tab"
            data-node-key="questions"
          >
            <div
              aria-controls="rc-tabs-test-panel-questions"
              aria-selected="false"
              class="ant-tabs-tab-btn"
              id="rc-tabs-test-tab-questions"
              role="tab"
              tabindex="-1"
            >
              Questions (0)
            </div>
          </div>
          <div
            class="ant-tabs-ink-bar ant-tabs-ink-bar-animated"
          />
        </div>
      </div>
      <div
        class="ant-tabs-nav-operations ant-tabs-nav-operations-hidden"
      >
        <button
          aria-controls="rc-tabs-test-more-popup"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="ant-tabs-nav-more"
          id="rc-tabs-test-more"
          style="visibility: hidden; order: 1;"
          type="button"
        >
          <span
            aria-label="ellipsis"
            class="anticon anticon-ellipsis"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="ellipsis"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
    <div
      class="ant-tabs-content-holder"
    >
      <div
        class="ant-tabs-content ant-tabs-content-top"
      >
        <div
          aria-hidden="false"
          aria-labelledby="rc-tabs-test-tab-basic"
          class="ant-tabs-tabpane ant-tabs-tabpane-active"
          id="rc-tabs-test-panel-basic"
          role="tabpanel"
          tabindex="0"
        >
          <div
            class="ant-card ant-card-bordered css-dev-only-do-not-override-1vjf2v5"
          >
            <div
              class="ant-card-body"
            >
              <form
                class="ant-form ant-form-vertical css-dev-only-do-not-override-1vjf2v5"
              >
                <div
                  class="ant-row css-dev-only-do-not-override-1vjf2v5"
                  style="margin-left: -8px; margin-right: -8px;"
                >
                  <div
                    class="ant-col ant-col-12 css-dev-only-do-not-override-1vjf2v5"
                    style="padding-left: 8px; padding-right: 8px;"
                  >
                    <div
                      class="ant-form-item css-dev-only-do-not-override-1vjf2v5"
                    >
                      <div
                        class="ant-row ant-form-item-row css-dev-only-do-not-override-1vjf2v5"
                      >
                        <div
                          class="ant-col ant-form-item-label css-dev-only-do-not-override-1vjf2v5"
                        >
                          <label
                            class="ant-form-item-required"
                            for="title"
                            title=""
                          >
                            <span />
                          </label>
                        </div>
                        <div
                          class="ant-col ant-form-item-control css-dev-only-do-not-override-1vjf2v5"
                        >
                          <div
                            class="ant-form-item-control-input"
                          >
                            <div
                              class="ant-form-item-control-input-content"
                            >
                              <input
                                aria-required="true"
                                class="ant-input css-dev-only-do-not-override-1vjf2v5 ant-input-outlined"
                                id="title"
                                placeholder="Enter quiz title"
                                type="text"
                                value=""
                              />
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ant-col ant-col-12 css-dev-only-do-not-override-1vjf2v5"
                    style="padding-left: 8px; padding-right: 8px;"
                  >
                    <div
                      class="ant-form-item css-dev-only-do-not-override-1vjf2v5"
                    >
                      <div
                        class="ant-row ant-form-item-row css-dev-only-do-not-override-1vjf2v5"
                      >
                        <div
                          class="ant-col ant-form-item-label css-dev-only-do-not-override-1vjf2v5"
                        >
                          <label
                            class="ant-form-item-required"
                            for="quizType"
                            title=""
                          >
                            <span />
                          </label>
                        </div>
                        <div
                          class="ant-col ant-form-item-control css-dev-only-do-not-override-1vjf2v5"
                        >
                          <div
                            class="ant-form-item-control-input"
                          >
                            <div
                              class="ant-form-item-control-input-content"
                            >
                              <div
                                aria-required="true"
                                class="ant-select ant-select-outlined ant-select-in-form-item css-dev-only-do-not-override-1vjf2v5 ant-select-single ant-select-show-arrow"
                              >
                                <div
                                  class="ant-select-selector"
                                >
                                  <span
                                    class="ant-select-selection-wrap"
                                  >
                                    <span
                                      class="ant-select-selection-search"
                                    >
                                      <input
                                        aria-autocomplete="list"
                                        aria-controls="quizType_list"
                                        aria-expanded="false"
                                        aria-haspopup="listbox"
                                        aria-owns="quizType_list"
                                        aria-required="true"
                                        autocomplete="off"
                                        class="ant-select-selection-search-input"
                                        id="quizType"
                                        readonly=""
                                        role="combobox"
                                        style="opacity: 0;"
                                        type="search"
                                        unselectable="on"
                                        value=""
                                      />
                                    </span>
                                    <span
                                      class="ant-select-selection-item"
                                      title="Lesson Quiz"
                                    >
                                      Lesson Quiz
                                    </span>
                                  </span>
                                </div>
                                <span
                                  aria-hidden="true"
                                  class="ant-select-arrow"
                                  style="user-select: none;"
                                  unselectable="on"
                                >
                                  <span
                                    aria-label="down"
                                    class="anticon anticon-down ant-select-suffix"
                                    role="img"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      data-icon="down"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="64 64 896 896"
                                      width="1em"
                                    >
                                      <path
                                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                                      />
                                    </svg>
                                  </span>
                                </span>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-form-item css-dev-only-do-not-override-1vjf2v5"
                >
                  <div
                    class="ant-row ant-form-item-row css-dev-only-do-not-override-1vjf2v5"
                  >
                    <div
                      class="ant-col ant-form-item-label css-dev-only-do-not-override-1vjf2v5"
                    >
                      <label
                        class=""
                        for="description"
                        title=""
                      >
                        <span />
                      </label>
                    </div>
                    <div
                      class="ant-col ant-form-item-control css-dev-only-do-not-override-1vjf2v5"
                    >
                      <div
                        class="ant-form-item-control-input"
                      >
                        <div
                          class="ant-form-item-control-input-content"
                        >
                          <textarea
                            class="ant-input css-dev-only-do-not-override-1vjf2v5 ant-input-outlined"
                            id="description"
                            placeholder="Enter quiz description"
                            rows="3"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-row css-dev-only-do-not-override-1vjf2v5"
                  style="margin-left: -8px; margin-right: -8px;"
                >
                  <div
                    class="ant-col ant-col-8 css-dev-only-do-not-override-1vjf2v5"
                    style="padding-left: 8px; padding-right: 8px;"
                  >
                    <div
                      class="ant-form-item css-dev-only-do-not-override-1vjf2v5"
                    >
                      <div
                        class="ant-row ant-form-item-row css-dev-only-do-not-override-1vjf2v5"
                      >
                        <div
                          class="ant-col ant-form-item-label css-dev-only-do-not-override-1vjf2v5"
                        >
                          <label
                            class=""
                            for="timeLimitMinutes"
                            title=""
                          >
                            <span />
                          </label>
                        </div>
                        <div
                          class="ant-col ant-form-item-control css-dev-only-do-not-override-1vjf2v5"
                        >
                          <div
                            class="ant-form-item-control-input"
                          >
                            <div
                              class="ant-form-item-control-input-content"
                            >
                              <div
                                class="ant-input-number ant-input-number-in-form-item css-dev-only-do-not-override-1vjf2v5 ant-input-number-outlined"
                                style="width: 100%;"
                              >
                                <div
                                  class="ant-input-number-handler-wrap"
                                >
                                  <span
                                    aria-disabled="false"
                                    aria-label="Increase Value"
                                    class="ant-input-number-handler ant-input-number-handler-up"
                                    role="button"
                                    unselectable="on"
                                  >
                                    <span
                                      aria-label="up"
                                      class="anticon anticon-up ant-input-number-handler-up-inner"
                                      role="img"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        data-icon="up"
                                        fill="currentColor"
                                        focusable="false"
                                        height="1em"
                                        viewBox="64 64 896 896"
                                        width="1em"
                                      >
                                        <path
                                          d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                                        />
                                      </svg>
                                    </span>
                                  </span>
                                  <span
                                    aria-disabled="false"
                                    aria-label="Decrease Value"
                                    class="ant-input-number-handler ant-input-number-handler-down"
                                    role="button"
                                    unselectable="on"
                                  >
                                    <span
                                      aria-label="down"
                                      class="anticon anticon-down ant-input-number-handler-down-inner"
                                      role="img"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        data-icon="down"
                                        fill="currentColor"
                                        focusable="false"
                                        height="1em"
                                        viewBox="64 64 896 896"
                                        width="1em"
                                      >
                                        <path
                                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                                        />
                                      </svg>
                                    </span>
                                  </span>
                                </div>
                                <div
                                  class="ant-input-number-input-wrap"
                                >
                                  <input
                                    aria-valuemax="300"
                                    aria-valuemin="1"
                                    autocomplete="off"
                                    class="ant-input-number-input"
                                    id="timeLimitMinutes"
                                    placeholder="No limit"
                                    role="spinbutton"
                                    step="1"
                                    value=""
                                  />
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ant-col ant-col-8 css-dev-only-do-not-override-1vjf2v5"
                    style="padding-left: 8px; padding-right: 8px;"
                  >
                    <div
                      class="ant-form-item css-dev-only-do-not-override-1vjf2v5"
                    >
                      <div
                        class="ant-row ant-form-item-row css-dev-only-do-not-override-1vjf2v5"
                      >
                        <div
                          class="ant-col ant-form-item-label css-dev-only-do-not-override-1vjf2v5"
                        >
                          <label
                            class=""
                            for="isTest"
                            title=""
                          >
                            <span />
                          </label>
                        </div>
                        <div
                          class="ant-col ant-form-item-control css-dev-only-do-not-override-1vjf2v5"
                        >
                          <div
                            class="ant-form-item-control-input"
                          >
                            <div
                              class="ant-form-item-control-input-content"
                            >
                              <button
                                aria-checked="false"
                                class="ant-switch css-dev-only-do-not-override-1vjf2v5"
                                id="isTest"
                                role="switch"
                                type="button"
                              >
                                <div
                                  class="ant-switch-handle"
                                />
                                <span
                                  class="ant-switch-inner"
                                >
                                  <span
                                    class="ant-switch-inner-checked"
                                  />
                                  <span
                                    class="ant-switch-inner-unchecked"
                                  />
                                </span>
                              </button>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ant-col ant-col-8 css-dev-only-do-not-override-1vjf2v5"
                    style="padding-left: 8px; padding-right: 8px;"
                  >
                    <div
                      class="ant-form-item css-dev-only-do-not-override-1vjf2v5"
                    >
                      <div
                        class="ant-row ant-form-item-row css-dev-only-do-not-override-1vjf2v5"
                      >
                        <div
                          class="ant-col ant-form-item-label css-dev-only-do-not-override-1vjf2v5"
                        >
                          <label
                            class=""
                            for="isPractice"
                            title=""
                          >
                            <span />
                          </label>
                        </div>
                        <div
                          class="ant-col ant-form-item-control css-dev-only-do-not-override-1vjf2v5"
                        >
                          <div
                            class="ant-form-item-control-input"
                          >
                            <div
                              class="ant-form-item-control-input-content"
                            >
                              <button
                                aria-checked="true"
                                class="ant-switch css-dev-only-do-not-override-1vjf2v5 ant-switch-checked"
                                id="isPractice"
                                role="switch"
                                type="button"
                              >
                                <div
                                  class="ant-switch-handle"
                                />
                                <span
                                  class="ant-switch-inner"
                                >
                                  <span
                                    class="ant-switch-inner-checked"
                                  />
                                  <span
                                    class="ant-switch-inner-unchecked"
                                  />
                                </span>
                              </button>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
